keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,448 @@
+
+#### Cardinality growth of metrics
+
+The proposal will add O(pods*resource_types) series. Currently a larger number of series are exposed via `kube-state-metrics` to help summarize to this data, but this proposal should reduce the need for those metrics and lead to a cardinality reduction in the future. This proposal grows the number of per pod series by a small fraction of the existing amount, described in the [kube-state-metrics project docs](https://github.com/kubernetes/kube-state-metrics/blob/master/docs/pod-metrics.md). In extremely large clusters an administrator may wish to not scrape or gather the metrics described here using the described flag.
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Expose new metrics
+
+The scheduler will expose an optional set of metrics series that capture the resources a pod is considered to be "consuming" from a scheduling perspective at the current time. The metrics will be exposed via `/metrics/resources` to allow aggregation to be optional and to allow a separate scrape interval. Only the active scheduler will report metrics series in order to remove the need to downsample the series and to reduce overall cardinality. If no scheduler is currently leading, then these metrics will temporarily not be reported which is consistent with metrics based on populated caches in leader elected components.
+
+The series reported will be consistent with `kube_node_status_allocatable` to simplify queries that compare used capacity and available capacity. After a pod has reached the Completed lifecycle (as described above) state, the series will not be reported. If the value of the resource is zero, that resource will not report a series. They all share the form `kube_(running_|)pod_resource_(requests|limits)`.
+
+Note on naming: The Prometheus convention for a metric is to allow the sum of a metric to make sense. However, in the case of Kubernetes resources we are attempting to make discoverability of the reported metrics (which are unbounded in number as they are extensible) a key goal for admins. Most admins will understand CPU and memory usage, but would have to resort to metrics name wildcard regexes to find the other values if we give each resource its own metrics name. Also, because Kube extension resources have characters disallowed in metrics names, admins would continually have to transform `myresource.io/foo-bar` into `myresource_io_foo_bar` in order to locate names which places an undue burden on the user. So for the proposed metrics and `kube_node_status_allocatable` we instead place all types resources under the same metrics name with a unit and resource label.
```

Comment:
I realize we agreed in the meeting to use a metric with a unit, so i'm not trying to re-litigate that.  For the "usage" metrics below, we split out CPU, Memory from extended resources.  Since the goal is to make the request/limit metrics easily join-able with the usage metrics, I would think we want to do either all generic or all per-resource (with extended resources grouped into one metric).

Reply:
The existing metrics today for usage are all "per resource" (when it comes to usage, not allocatable) whereas allocatable is generic and pod resources are generic.  I think the scope of change to move all usage metrics might be high, but we could start in 1.20 with defining what consistency is without requiring change (the new generic metric could be additive).